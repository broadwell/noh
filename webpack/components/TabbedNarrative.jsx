/* eslint-disable react/no-array-index-key */

import React from "react";
import PropTypes from "prop-types";
import { Markup } from "interweave";
import { Tab, Tabs, TabList, TabPanel } from "react-tabs";

class TabbedNarrative extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      narrative: props.narrative
    };
  }

  componentWillMount() {
    this.parseNarrative();
  }

  parseNarrative() {
    // return a list of strings broken on <br>
    const chunks = this.state.narrative.split("<br>");
    return chunks;
  }

  panelTitles() {
    // does this handle multi word titles?
    const chunks = this.parseNarrative();
    const re = /title='[\w\s]+'/;
    const titles = [];
    for (let i = 0; i < chunks.length; i += 1) {
      const titleString = re.exec(chunks[i]);
      const pieces = titleString[0].split("=");
      const title = pieces[1].replace("'", "").replace("'", "");
      titles.push(title);
    }
    return titles;
  }

  render() {
    const chunks = this.parseNarrative();
    const titles = this.panelTitles();
    const narrativeTabList = titles.map(title => (
      <Tab key={title}>{title}</Tab>
    ));
    const narrativeTabs = chunks.map((chunk, index) => (
      <TabPanel key={index}>
        <Markup content={chunk} />
      </TabPanel>
    ));
    return (
      <Tabs>
        <TabList>{narrativeTabList}</TabList>
        {narrativeTabs}
      </Tabs>
    );
  }
}

TabbedNarrative.propTypes = {
  narrative: PropTypes.string
};

TabbedNarrative.defaultProps = {
  narrative: ""
};

export default TabbedNarrative;
