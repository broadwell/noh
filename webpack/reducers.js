import { SET_CURRENT_TIME, SET_IS_PLAYING } from './actions';

const DEFAULT_STATE = {
  currentTime: 0,
  isPlaying: false
};

const setCurrentTime = (state, action) =>
  Object.assign({}, state, { currentTime: action.payload });

const setIsPlaying = (state, action) =>
  Object.assign({}, state, { isPlaying: action.payload });

const rootReducer = (state = DEFAULT_STATE, action) => {
  switch (action.type) {
    case SET_CURRENT_TIME:
      return setCurrentTime(state, action);
    case SET_IS_PLAYING:
      return setIsPlaying(state, action);
    default:
      return state;
  }
};

export default rootReducer;
