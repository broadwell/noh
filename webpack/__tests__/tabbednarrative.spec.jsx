import React from "react";
import { shallow, mount } from "enzyme";
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";
import { DEFAULT_STATE } from "../reducers";
import TabbedNarrative, {
  Unwrapped as UnwrappedTabbedNarrative
} from "../components/TabbedNarrative";

describe("<TabbedNarrative>", () => {
  const tabbedAnalysis =
    "<section title='section one with many words in title' class='tabbed-narrative'><p>Lorem ipsum dolor sit amet.</p></section><br /><section title='section two'><p>Different stuff</p></section>";

  it("renders as expected", () => {
    const component = shallow(
      <UnwrappedTabbedNarrative
        narrative={tabbedAnalysis}
        narrativeTab={0}
        updateNarrativeTab={() => jest.fn()}
      />
    );
    expect(component).toMatchSnapshot();
  });

  it("ref handler runs with tabs ref", () => {
    const component = mount(
      <div className="narrative" role="presentation">
        <UnwrappedTabbedNarrative
          narrative={tabbedAnalysis}
          narrativeTab={0}
          updateNarrativeTab={() => jest.fn()}
        />
      </div>
    );
    const tabN = component.find("TabbedNarrative");
    tabN.instance().handleDomRef(tabN.getDOMNode());
  });

  it("ref handler runs with null input", () => {
    const component = shallow(
      <UnwrappedTabbedNarrative
        narrative={tabbedAnalysis}
        narrativeTab={0}
        updateNarrativeTab={() => jest.fn()}
      />
    );
    component.instance().handleDomRef(null);
  });

  it("updates store correctly on tab change", () => {
    const mockStore = configureMockStore();
    const defaultState = DEFAULT_STATE;
    const store = mockStore(defaultState);
    const wrapper = mount(
      <Provider store={store}>
        <TabbedNarrative
          narrative={tabbedAnalysis}
          narrativeTab={0}
          updateNarrativeTab={() => jest.fn()}
        />
      </Provider>
    );

    const action = { type: "SET_NARRATIVE_TAB", payload: 1 };
    const button = wrapper.find("li.react-tabs__tab").last();
    button.simulate("click");
    expect(store.getActions()[0]).toEqual(action);
  });
});
