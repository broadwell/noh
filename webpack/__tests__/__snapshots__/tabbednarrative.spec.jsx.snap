// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TabbedNarrative> renders as expected 1`] = `
<Tabs
  defaultFocus={false}
  defaultIndex={null}
  domRef={[Function]}
  forceRenderTabPanel={false}
  selectedIndex={null}
>
  <TabList
    className="react-tabs__tab-list"
  />
</Tabs>
`;
